packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerParams.ts

```diff
@@ -95,15 +105,102 @@ class LayerState implements LayerMetadata {
   getHash(version: number): string {
     return this.getVersion(version).hash;
   }
+
+  // sets the hash for a new (not yet pushed) layer version
+  // once a hash is set for a version, this should indicate that the version is "finialized" (ie pushed or about to be pushed)
+  // hashes are immutable once set
+  async setNewVersionHash() {
+    const latestVersion = this.getLatestVersion();
+    // if the latest version doesn't already have a hash
+    if (!this.getHash(latestVersion)) {
+      const newHash = this.newVersionHash || (await this.hashLayer());
+      this.getVersion(latestVersion).hash = newHash;
+      this.storedParams.layerVersionMap[latestVersion].hash = newHash;
+      this.newVersionHash = undefined; // reset the newVersionHash now that the latest one is set
+    }
+  }
+
+  // updates the layer metadata with a new version if changes are detected
+  // if a new version is detected, the permissions from the previous version are carried forward to the new version
+  // returns true if a new version was detected, false otherwise
+  // it does not set a hash for a new version if detected because the version could change more before a push
+  async syncVersions(): Promise<boolean> {
+    const latestVersion = this.getLatestVersion();
+    const currHash = this.getHash(latestVersion);
+    if (!currHash) {
+      return false;
+    }
+    this.newVersionHash = await this.hashLayer();
+    if (currHash !== this.newVersionHash) {
+      this.addNewLayerVersion();
+      return true;
+    }
+    return false;
+  }
+
+  updateCompatibleRuntimes(runtimes: LayerRuntime[]) {
+    let isModified = runtimes.length !== this.runtimes.length;
+    const existingRuntimeVals = this.runtimes.map(runtime => runtime.value);
+    const newRuntimeVals = runtimes.map(runtime => runtime.value);
+    // check that the arrays are different
+    isModified =
+      isModified ||
+      existingRuntimeVals.findIndex(val => !newRuntimeVals.includes(val)) !== -1 ||
```

Comment:
Can we use  isModified = _.isEqual(existingRuntimeVals.sort() , newRuntimeVals.sort())

Reply:
good call, will change that